### ISSUE TITLE
How do I select the Pre CPU, hierarchical, and Ring modes when I use Adasum?

### ISSUE THREAD

**Issue by liuyunfeng2016** [horovod/horovod#2582]
@ashbhandare @tgaddair @sblotner hello,I found that hvd.AdaSum is incorrect in Horovod 0.21.0. It should be hvd.Adasum.
Then I have some questions when I use Adasum,how do I choose Pure CPU、Ring、Hierarchical、three modes
Now,I found HOROVOD_HIERARCHICAL_ALLREDUCE in paper,
my environment is 8 Tesla V100s per node.How can I make the most of his ability?
The following are my building parameters.

ENV HOROVOD_WITHOUT_GLOO=1
ENV HOROVOD_GPU_OPERATIONS='NCCL'
ENV HOROVOD_CPU_OPERATIONS='MPI'
ENV HOROVOD_WITH_PYTORCH=1
ENV HOROVOD_NCCL_INCLUDE=/usr/include
ENV HOROVOD_NCCL_LIB=/usr/lib/x86_64-linux-gnu
ENV HOROVOD_MPICXX_SHOW="/usr/local/openmpi/bin/mpicxx -show"
ENV HOROVOD_WITH_TENSORFLOW=1
ENV HOROVOD_CUDA_HOME="/usr/local/cuda"
ENV TENSORFLOW_VERSION=1.15

**Comment by chongxiaoc** [comment]
See the code here:

https://github.com/horovod/horovod/blob/a9dea74abc1f0b8e81cd2b6dd9fe81e2c4244e39/horovod/tensorflow/__init__.py#L125

- tensor is CPU:
  - Pure CPU: number of ranks have to be power of 2 
- tensor is GPU:
  - if `nccl` is enabled:
    - GPU adasum is used only if cluster is  `homogeneous` (every node has same amount of ranks) and local size has to be power of 2.
   - if `nccl` is not enabled:
     -    Pure CPU

I read above from code, please correct me if I'm wrong.

**Comment by liuyunfeng2016** [comment]
> See the code here:
> 
> https://github.com/horovod/horovod/blob/a9dea74abc1f0b8e81cd2b6dd9fe81e2c4244e39/horovod/tensorflow/__init__.py#L125
> 
> * tensor is CPU:
>   
>   * Pure CPU: number of ranks have to be power of 2
> * tensor is GPU:
>   
>   * if `nccl` is enabled:
>     
>     * GPU adasum is used only if cluster is  `homogeneous` (every node has same amount of ranks) and local size has to be power of 2.
>   * if `nccl` is not enabled:
>     
>     * Pure CPU
> 
> I read above from code, please correct me if I'm wrong.


Thank you very much for your help, but it seems that "summed_tensor" has completed calculations in the above function,
https://github.com/horovod/horovod/blob/a9dea74abc1f0b8e81cd2b6dd9fe81e2c4244e39/horovod/tensorflow/__init__.py#L120
I read this [docs](https://horovod.readthedocs.io/en/latest/adasum_user_guide_include.html) with adasum,but can't find solutions

**Comment by chongxiaoc** [comment]
I'm not a pro on tensorflow side, but pytorch implementation is as I expected, all_reduce returns an async handle for adasum:
https://github.com/horovod/horovod/blob/a9dea74abc1f0b8e81cd2b6dd9fe81e2c4244e39/horovod/torch/mpi_ops.py#L129

**Comment by liuyunfeng2016** [comment]
> I'm not a pro on tensorflow side, but pytorch implementation is as I expected, all_reduce returns an async handle for adasum:
> https://github.com/horovod/horovod/blob/a9dea74abc1f0b8e81cd2b6dd9fe81e2c4244e39/horovod/torch/mpi_ops.py#L129

However, I still don't know how to modify the contents of the dynamic library or the variables to select the three methods of Adasum.
Does this file modify the dynamic library? But I didn't find the relevant environment variables.
https://github.com/horovod/horovod/blob/master/horovod/common/ops/adasum/adasum_mpi.cc#L16

**Comment by tgaddair** [comment]
Hey @liuyunfeng2016, `HOROVOD_HIERARCHICAL_ALLREDUCE` is deprecated for recent versions of NCCL, which have this capability build in automatically.  Adasum will not work with hierarchical allreduce, as it uses its own aggregation method.  It also does not use NCCL for cross-host communication at this time, though there are plans to update it to NCCL now that NCCL 2.7 supports point-to-point communication.

So with Adasum, to maximize performance you'll want to NCCL and scale the learning rate by `hvd.local_size()` (in your case: 8).

**Comment by liuyunfeng2016** [comment]
> Hey @liuyunfeng2016, `HOROVOD_HIERARCHICAL_ALLREDUCE` is deprecated for recent versions of NCCL, which have this capability build in automatically. Adasum will not work with hierarchical allreduce, as it uses its own aggregation method. It also does not use NCCL for cross-host communication at this time, though there are plans to update it to NCCL now that NCCL 2.7 supports point-to-point communication.
> 
> So with Adasum, to maximize performance you'll want to NCCL and scale the learning rate by `hvd.local_size()` (in your case: 8).

I see. In ring mode, the precision is the best, but the intra-node is the Adasum of the MPI, so it performance is not the best. In hierarchical mode, the performance is the best. However, because the intra-node is implemented by the NCCL with Averaging, the precision is not the best but the fastest. Therefore, you need to increase the learning rate of the number of cards on the node to reduce the precision loss. When I set ENV HOROVOD_GPU_OPERATIONS= 'NCCL', it indicates that the device is in hierarchical mode.

**Comment by tgaddair** [comment]
Hey @liuyunfeng2016, to answer your questions:

- Increasing batches per allreduce will decrease network utilization by increasing effective batch size.
- Gradient predivide factor can be tuned based empirically if you're seeing a trend towards overflow / underflow (cc @romerojosh ).
- FP16 allreduce we decrease network utilization by reducing precision, which can lead to overflow / underflow.
- You don't need to use the `DistributedDataParallel` or `init_process_group` parts for Horovod, as these are handled by `DistributedOptimizer` and `hvd.init()`.

**Comment by liuyunfeng2016** [comment]
@tgaddair   OK，I see. Thank you very much for your answer.
